### PR TITLE
Fix unknown inputs deserialization

### PR DESCRIPTION
**Pull request by jkerken**
Fix extracted from https://github.com/pulumi/pulumi-dotnet/pull/275

I also found the same bug for unkown secrets inputs. Which is also fixed.

I did not copy the intgeration test over since that was done using a c# component provider
But it is present in  the other pr, I added some unit tests to cover these edge cases though.

### CODE REVIEW

File: .changes/unreleased/Bug Fixes-304.yaml [hunk 1]
```diff
@@ -1,6 +1,6 @@
 component: sdk
 kind: Bug Fixes
-body: Fix race condition in GrpcMonitor's GrpcChannel management
```
Comment (Frassle) [inline .changes/unreleased/Bug Fixes-304.yaml:3]: Can leave this file as is

File: sdk/Pulumi.Tests/Provider/PropertyValueTests.cs [hunk 2]
```diff
@@ -299,6 +315,24 @@ public async Task DeserializingWrappedOutputSecretWorks()
         Assert.True(data.IsKnown);
     }
 
+    [Fact]
+    public async Task DeserializingWrappedUnknownOutputSecretWorks()
+    {
+        var serializer = CreateSerializer();
+        // secretOutput = Secret(Output(Unknown))
```
Comment (Frassle) [inline sdk/Pulumi.Tests/Provider/PropertyValueTests.cs:322]: ?

File: sdk/Pulumi.Tests/Provider/PropertyValueTests.cs [hunk 3]
```diff
@@ -299,6 +315,24 @@ public async Task DeserializingWrappedOutputSecretWorks()
...
+        var secretOutput = new PropertyValue(
+            new PropertyValue(new OutputReference(
+                value:  PropertyValue.Computed,
+                dependencies: ImmutableArray<Urn>.Empty)));
```
Comment (Frassle) [inline sdk/Pulumi.Tests/Provider/PropertyValueTests.cs:326]: Shouldn't "secret" be on the OutputReference object?

File: .changes/unreleased/Bug Fixes-306.yaml [hunk 4]
```diff
@@ -0,0 +1,6 @@
+component: sdk
+kind: Bug Fixes
+body: Fix unkown inputs deserialization
```
Comment (Frassle) [inline .changes/unreleased/Bug Fixes-306.yaml:3]:
```suggestion
body: Fix unknown inputs deserialization
```

File: sdk/Pulumi.Tests/Provider/PropertyValueTests.cs (hunk 3 above)
Comment (Frassle) [reply sdk/Pulumi.Tests/Provider/PropertyValueTests.cs:326]: Ah no, this just needs to wrap in the double PropertyValue as it's doing.

Comment (jkerken) [reply sdk/Pulumi.Tests/Provider/PropertyValueTests.cs:326]: In deserialize there are both cases covered.

- Secret(Output(...)) being deserialized into Input<T>
- Output(Secret(...)) being deserialized into Input<T>

There is also already: 

- DeserializingWrappedSecretInOutputWorks which deserializes Output(Secret("Hello"))
- DeserializingWrappedOutputSecretWorks which deserializes Secret(Output("Hello"))

I just assumed that both value can be unknown

File: .changes/unreleased/Bug Fixes-306.yaml (hunk 4 above)
Comment (jkerken) [reply .changes/unreleased/Bug Fixes-306.yaml]: Fixed

File: sdk/Pulumi.Tests/Provider/PropertyValueTests.cs (hunk 3 above)
Comment (Frassle) [reply sdk/Pulumi.Tests/Provider/PropertyValueTests.cs:326]: Yeh either case could be unknown.

File: sdk/Pulumi.Tests/Provider/PropertyValueTests.cs (hunk 2 above)
Comment (Frassle) [inline sdk/Pulumi.Tests/Provider/PropertyValueTests.cs:322]:
```suggestion

```

File: sdk/Pulumi.Tests/Provider/PropertyValueTests.cs [hunk 5]
```diff
@@ -317,6 +351,25 @@ public async Task DeserializingWrappedSecretInOutputWorks()
         Assert.True(data.IsKnown);
     }
 
+
+    [Fact]
+    public async Task DeserializingWrappedUnknownSecretInOutputWorks()
+    {
+        var serializer = CreateSerializer();
+        // secretOutput = Output(Secret(Unknown))
```
Comment (Frassle) [inline sdk/Pulumi.Tests/Provider/PropertyValueTests.cs:359]: I think these comments caused me more confusion than understanding, I'd just take them out. If you want a description here put it in English "secretOutput is Unknown wrapped in a Secret wrapped in an Output"

**Review by Zaid-Ajaj — APPROVED** [review]
